Various Fixes and Improves

kolbyb opened the pull request:
- Fixed some of the Perspective Issues
  > Merged our planar approximation changes.
  >
  > Adjusted the `Segment` class so that `direction` is a cached property.
  > Constructing a `Ray` from `Segment` no longer requires `atan2`, instead favoring `Segment.direction`.
  >
  > Added a fix for a bug introduced by the new the ray generation process in `Segment.rays(...)`.
  > _When the Camera is axis-aligned, a ray in the center of the view will pass through walls._
  > _I've added an epsilon/bias so that no axis-aligned rays should be generated._
- Fixed Movement
  > Moving Forward/Backward and rotating Left/Right is now based off frame time.
- Added a 2d map display primarily for debugging purposes
  > This is on by default but can be changed by editing _Line 549_
- Added debug options
  > Press **M** to toggle the 2d map
  > Press **P** to toggle the perspective approximation
  > Press **C** to toggle an alternate coloring mode based on surface normals
  > Press **R** to toggle Rays being drawn on the 2d map
- Introduced the concept of segment (surface) normals
  > Useful for calculating various things, including lighting approximations.
  >
  > This simply uses a fixed rotation from the `Segment.normal`. It's crude, but would need to be taken into account during map generation to invert generated segments, to address inverted normals. Alternatively, this could be an additional generated property.

kolbyb committed:
- Fixed some of the perspective issues
- Fixed movement not being time based
- Added a 2d map display primarily for debugging purposes
- Added debug options
  Press M to toggle the 2d map
  Press P to toggle the perspective approximation
  Press C to toggle an alternate coloring mode based on surface normals
  Press R to toggle Rays being drawn on the 2d map
- Introduced the concept of segment (surface) normals

kolbyb committed:
Merge branch 'main' of https://github.com/kolbyb/raycasting

kolbyb committed:
- Merged to Main

kolbyb committed:
- Add WASD + Mouse Controls
- Add a direction helper function to Point

lefticus commented:
Unfortunately I cannot accept this PR, as much of what you have done are things I have planned for future episodes and streaming and are things I want to add myself!

I am interested in these fixes, as they are, specifically, fixes:

> Fixed some of the Perspective Issues
>
>   Merged our planar approximation changes.
>
>    Adjusted the Segment class so that direction is a cached property.
>    Constructing a Ray from Segment no longer requires atan2, instead favoring Segment.direction.
>
>    Added a fix for a bug introduced by the new the ray generation process in Segment.rays(...).
>    When the Camera is axis-aligned, a ray in the center of the view will pass through walls.
>    I've added an epsilon/bias so that no axis-aligned rays should be generated.

If you are interested in splitting that off into a separate PR

kolbyb committed:
- Fix the point rotation for calculating segment
  forward (surface normals).
- Update the alternate coloring to show what a segment
  surface normal is with lighting based on if the camera
  is facing the surface.

kolbyb commented:
Absolutely fair. I can spit the smaller fixes into specific CLs/PRs after I look back over what was changed since your latest stream 😄

lefticus commented:
The more I spent time on those tests today the more I want to know *why* that perpendicular ray passes through the walls, and the less I want to try and work around it. We'll definitely need to keep working on that test suite :D
